src/cocaine-app/sched/__init__.py

```diff
@@ -0,0 +1,417 @@
+                logger.info("No sense to cancel cause of res={} (demand={} < usage_if_cancel={}, jobs are {})".format(
+                    res, res_demand, usage_if_cancel, self.res[res]))
+                return False
+
+        # XXX: Now we cancel all jobs while we may be satisfied with less amount. Fix that
+
+        try:
+            self.job_processor.stop_jobs_list(filter(lambda x: x.id in cancellable_jobs, existing_jobs))
+        except:
+            logger.exception("Failed to cancel jobs {}".format(cancellable_jobs))
+            return False
+
+        logger.info("Successfully cancelled {}".format(cancellable_jobs))
+        return True
+
+    def process_jobs(self, job_type, jobs_param_list, sched_params, filter_func=None):
+        """
+        Scheduler creates jobs with specified params of specified type. Sched_params and filter_func
+        :param job_type: job type
+        :param jobs_param_list: a list of dictionaries with params for jobs created.
+                                A number of job created <= len(jobs_param_list)
+        :param sched_params: in a common case sched params are stored in config in "scheduler/{starter}" section.
+                            But in certain cases the starter may want to redefine some parameters.
+                            For example, one may want to increase priority or decrease max_deadline_time
+        :param filter_func: Filter function are called before job creation to extra-check whether those jobs are needed
+                            See more details on wiki
+                            (https://wiki.yandex-team.ru/users/aniam/mm-scheduler/#generacijadzhobovstartdzhobov)
+        :return: a number of generated jobs
+        """
+
+        self.update_resource_stat()
+
+        max_jobs = sched_params.get('max_executing_jobs')
+        if max_jobs:
+            job_count = self.job_count[job_type]
+            if job_count >= max_jobs:
+                logger.info('Found {} unfinished jobs (>= {}) of {} type'.format(job_count, max_jobs, job_type))
+                return 0
+            max_jobs -= job_count
+
+        created_job = 0
+
+        # common param for all types of jobs
+        need_approving = not sched_params.get('autoapprove', False)
+        force = sched_params.get('force', False)
+
+        job_class = JobFactory.make_job_type(job_type)
+
+        if not hasattr(job_class, 'report_resources'):
+            logger.error("Couldn't schedule the job {}. Add static report_resources function".format(job_type))
+            return 0
+
+        job_report_resources = job_class.report_resources
+
+        for job_param in jobs_param_list:
+
+            # Get resource demand and verify whether run is possible
+            old_res = job_report_resources(job_param)
+            res = self.convert_resource_representation(old_res['resources'], old_res['groups'], job_type)
+
+            crossing_jobs = self.get_resource_crossing(res)
```

Comment:
Done in https://github.com/yandex/mastermind/pull/417/commits/e7a0f53468278b6dc781abea9c82cb3cb7350f4e